### PR TITLE
add caching utils

### PR DESCRIPTION
**Pull request by joelgrus**
these fileutils will (once they're integrated elsewhere) allow us to specify URLs instead of files, resolving them to a local cache and downloading only if necessary.

this will allow us to point default configs to datasets / glove vectors on s3 instead of fixed paths.

we can also use this with `run evaluate` and so on

### CODE REVIEW

File: allennlp/common/file_utils.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    if parsed.scheme in ('http', 'https'):
+        # URL, so get it from the cache (downloading if necessary)
+        return get_from_cache(url_or_filename, cache_dir)
+    elif parsed.scheme == '' and os.path.exists(url_or_filename):
+        # File, and it exists.
+        return url_or_filename
+    elif parsed.scheme == '':
+        # File, but it doesn't exist.
+        raise FileNotFoundError("file {} not found".format(url_or_filename))
+    else:
```
Comment (matt-gardner) [inline allennlp/common/file_utils.py]: Is it worth checking for things like `file://`?  (I'd say probably not, just checking.)

File: tests/common/file_utils_test.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from allennlp.common.testing import AllenNlpTestCase
+
+
+class TestFileUtils(AllenNlpTestCase):
+
+    def test_url_to_filename(self):
+        for url in ['http://allenai.org', 'http://allennlp.org',
+                    'https://www.google.com', 'http://pytorch.org']:
+            filename = url_to_filename(url)
+            assert "http" not in filename
```
Comment (matt-gardner) [inline tests/common/file_utils_test.py]: This is probably good enough, but if you want to _actually_ check that you have something that's a valid filename, you could touch `os.path.join(self.TEST_DIR, filename)`.

File: allennlp/common/file_utils.py (hunk 1 above)
Comment (joelgrus) [inline allennlp/common/file_utils.py]: I considered it, but it made the code messier for no real benefit, let's leave it out for now

File: tests/common/file_utils_test.py (hunk 2 above)
Comment (joelgrus) [inline tests/common/file_utils_test.py]: I mostly added that to make sure that `url_to_filename` wasn't just leaving it as is, but touch is a good idea for a test, so I added it

**Pull request merged**